docs/decisions/0007-javascript-file-configuration.rst

```diff
@@ -0,0 +1,111 @@
+Promote JavaScript file configuration and deprecate environment variable configuration
+======================================================================================
+
+Status
+------
+
+Draft
+
+Context
+-------
+
+Our webpack build process allows us to set environment variables on the command line or via .env
+files.  These environment variables are available in the application via ``process.env``.
+
+The implementation of this uses templatization and string interpolation to replace any instance of
+``process.env.XXXX`` with the value of the environment variable named ``XXXX``.  As an example, in our
+source code we may write::
+
+    const LMS_BASE_URL = process.env.LMS_BASE_URL;
+
+After the build process runs, the compiled source code will instead read::
+
+    const LMS_BASE_URL = 'http://localhost:18000';
+
+Put another way, `process.env` is not actually an object available at runtime, it's a templatization
+token that helps the build replace it with a string literal.
+
+This approach has several important limitations:
+
+- There's no way to add variables without hard-coding process.env.XXXX somewhere in the file,
+  complicating our ability to add additional application-specific configuration without explicitly
+  merging it into the configuration document after it's been created in frontend-platform.
+- The method can *only* handle strings.
+
+  Other data types are converted to strings::
+
+    # Build command:
+    BOOLEAN_VAR=false NULL_VAR=null NUMBER_VAR=123 npm run build
+
+    ...
+
+    // Source code:
+    const BOOLEAN_VAR = process.env.BOOLEAN_VAR;
+    const NULL_VAR = process.env.NULL_VAR;
+    const NUMBER_VAR = process.env.NUMBER_VAR;
+
+    ...
+
+    // Compiled source after the build runs:
+    const BOOLEAN_VAR = "false";
+    const NULL_VAR = "null";
+    const NUMBER_VAR = "123";
+
+  This is not good!
+
+- It makes it very difficult to supply array and object configuration variables, and unreasonable to
+  supply class or function config since we'd have to ``eval()`` them.
+
+Related to all this, frontend-platform has long hand the ability to replace the implementations of
+its analytics, auth, and logging services, but no way to actually *configure* the app with a new
+implementation.  Because of the above limitations, there's no reasonable way to configure a
+JavaScript class via environment variables.
+
+Decision
+--------
+
+For the above reasons, we will deprecate environment variable configuration in favor of JavaScript
+file configuration.
```

Comment:
We might want to keep the legacy environment variable configuration around and not deprecate it. There's not much reason to deprecate it that I can think of (e.g., not much maintenance overhead to support it), though it'll probably start to get used less with the introduction of `env.config` and the MFE runtime config API.

We could observe usage over time and, if we see that very few MFEs use the environment variable configuration mechanism then, deprecate at that time.

Reply:
I hear you - the thing that comes to mind is that if we start using env.config.js so that we can pass real data types into the app instead of just strings, we'll presumably start removing the code that converts `"false"` to `false`, etc., in which case for given MFEs the old config won't work properly anymore.  Unless we just make the logic _more_ complicated to accommodate both. 😬  I guess this is a problem even in the interim before an official deprecation/removal.  Hm - will have to think about this more.

Reply:
On second thought, it may not be too much of a problem.  The MFEs tend to have arguments in their calls to `initialize` where they pull config out of `process.env` if it exists.  The new `jsFileConfig` function runs _after_ that.  The first can continue to massage strings into other data types, and the second doesn't have to worry about it.  They can co-exist.